src/EditorFeatures/TestUtilities/Completion/AbstractCompletionProviderTests.cs

```diff
@@ -246,7 +316,8 @@ private async Task VerifyAsync(
             string displayTextPrefix, string inlineDescription, bool? isComplexTextEdit,
             List<CompletionFilter> matchingFilters, CompletionItemFlags? flags, CompletionOptions options, bool skipSpeculation = false)
         {
-            foreach (var sourceKind in sourceCodeKind.HasValue ? [sourceCodeKind.Value] : new[] { SourceCodeKind.Regular, SourceCodeKind.Script })
+            IReadOnlyList<SourceCodeKind> evaluatedSourceCodeKinds = sourceCodeKind.HasValue ? [sourceCodeKind.Value] : [SourceCodeKind.Regular, SourceCodeKind.Script];
```

Comment:
If we are talking about perf, it makes sense to use a concrete type here (like a `List` or an array) to avoid allocating an enumerator. Given how many times this is invoked, it will probably be observable

Reply:
I doubt the extra enumerator allocation will be observable in any way, those tests are heavy on their own and this would be too tiny to affect the actual perf of the test. Nonetheless, reducing the `foreach` into a simple `for` in the runtime is a good goal. Iirc, the compiler reduces collection expressions with a fixed count returning `IReadOnlyList<...>` to a custom type accomodating the fixed number of elements. Will resort to an array for the time being.

Reply:
> the compiler reduces collection expressions with a fixed count returning `IReadOnlyList<...>` to a custom type accomodating the fixed number of elements

The reliance on implementation detail is not a good thing to begin with. Usually, if a concrete type can be used, it should be used instead of an abstraction

> this would be too tiny to affect the actual perf of the test

I won't argue on this. But given that type change here is pretty much free (since you immediately consume the collection), why not to go with it?

> Will resort to an array for the time being

SGTM